Scheduler simplified MultiPoint plugin config

damemi opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This adds a `MultiPoint` scheduler config field. This field is to be used by administrators to simplify the necessary config for most basic scheduler setups.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #102303

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
KubeSchedulerConfiguration provides a new field `MultiPoint` which will register a plugin for all valid extension points
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig scheduling

damemi committed:
Add MultiPoint scheduler plugin config field

alculquicondor reviewed (commented):
The changes in runtime/framework.go need tests

alculquicondor reviewed (commented):
good for squash IMO

Then, @ahg-g wanted to take a look

damemi commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -352,6 +336,60 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
+			}
+
+			if enabledSet.Has(ep.Name) {
+				return nil, fmt.Errorf("plugin %q already registered as %q", ep.Name, pluginType.Name())
```
It can be enabled, but you have to disable it first. That's the current behavior if you want to modify default weight right? Or did we change that?

alculquicondor replied:
We changed that

ahg-g replied:
ok, lets keep it that way then for Score.

damemi replied:
true, Score is the only case where it matters. Do you have a link to the changes? I remember talking about that issue with disabling/reenabling for weights but can't find the PR.

damemi replied:
Updated this, I realized that we only need to check double-registration for MultiPoint plugins against other MultiPoint plugins. This is because we are choosing to give explicit, individual extensions precedence. So by keeping track of the individual plugins (already registered above this point) separate from MultiPoint, we can handle that error by just skipping, exactly how we do with the default plugins now.

damemi commented on pkg/scheduler/profile/profile.go:
```diff
@@ -86,20 +85,25 @@ type cfgValidator struct {
-		return fmt.Errorf("one queue sort plugin required for profile with scheduler name %q", cfg.SchedulerName)
+
+	queueSort := f.ListPlugins().QueueSort.Enabled[0].Name
+	if len(queueSort) == 0 {
```
tl;dr: I fixed the test case to preserve the plugin names

It's actually a conflict of two different tests based on what functions they involve. If we set this line to use queueSort from the config (ie, `queueSort := cfg.Plugins.QueueSort.Enabled[0].Name`), then `TestSetup` fails. This failure is because the default plugins now come from MultiPoint, so there are no `QueueSort` plugins in the config until they are expanded by the factory logic:
```
$ go test ./cmd/kube-scheduler/...
...
--- FAIL: TestSetup (0.78s)
    --- FAIL: TestSetup/default_config (0.78s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 100 [running]:
testing.tRunner.func1.2({0x2819100, 0xc00031e150})
	/usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x218
panic({0x2819100, 0xc00031e150})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
k8s.io/kubernetes/pkg/scheduler/profile.(*cfgValidator).validate(0xc000a6b670, {{0x294deb8, 0x11}, 0xc0004b38c0, {0xc000156540, 0x7, 0x7}}, {0x2c37768, 0xc00025c8c0})
	/Users/mdame/go/src/k8s.io/kubernetes/pkg/scheduler/profile/profile.go:103 +0x5b0
k8s.io/kubernetes/pkg/scheduler/profile.NewMap({0xc00044b8f0, 0x1, 0x0}, 0xffffffffffffffff, 0x0, {0xc000a6b9b8, 0xb, 0xb})
	/Users/mdame/go/src/k8s.io/kubernetes/pkg/scheduler/profile/profile.go:61 +0x1d9
k8s.io/kubernetes/pkg/scheduler.(*Configurator).create(0xc000a6bae0)
	/Users/mdame/go/src/k8s.io/kubernetes/pkg/scheduler/factory.go:143 +0xbd3
k8s.io/kubernetes/pkg/scheduler.New({0x2c396d0, 0xc0001d4f00}, {0x2c2e848, 0xc00076c140}, {0x2bd9ad0, 0xc00076c1e0}, 0xc00052c600, 0x0, {0xc000907d20, 0xa, ...})
	/Users/mdame/go/src/k8s.io/kubernetes/pkg/scheduler/scheduler.go:264 +0x47b
k8s.io/kubernetes/cmd/kube-scheduler/app.Setup({0x2be0f98, 0xc00004eec0}, 0xc0003c0780, {0x0, 0x0, 0x3b7dbe0})
	/Users/mdame/go/src/k8s.io/kubernetes/cmd/kube-scheduler/app/server.go:309 +0x5d8
k8s.io/kubernetes/cmd/kube-scheduler/app.TestSetup.func3(0xc000582b60)
	/Users/mdame/go/src/k8s.io/kubernetes/cmd/kube-scheduler/app/server_test.go:241 +0x347
testing.tRunner(0xc000582b60, 0xc0006c4678)
	/usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x35a
...
```

But, if we flip it and just check the plugins from the completed factory (ie, `queueSort := f.ListPlugins().QueueSort.Enabled[0].Name`), then `TestNewMap` fails because we don't get the expected error:
```
$ go test ./pkg/scheduler/...
...
--- FAIL: TestNewMap (0.00s)
    --- FAIL: TestNewMap/different_queue_sort (0.00s)
        profile_test.go:251: got error %!q(<nil>), want "different queue sort plugins"
    --- FAIL: TestNewMap/different_queue_sort_args (0.00s)
        profile_test.go:251: got error %!q(<nil>), want "different queue sort plugin args"
FAIL
FAIL	k8s.io/kubernetes/pkg/scheduler/profile	0.345s
```

This is the test that is causing the issue. I debugged this to find that `f.ListPlugins()` was returning a list of plugins, but they didn't have any registered names. So, looking at `updatePluginList` in `framework.go`, I realized that we are just appending the plugin object. I'm assuming we then use `p.Name()` later on, because all of these test plugins are registered with the same empty `Name()` function. 

I fixed this by adding a wrapper to the test cases to pass an actual plugin name to each plugin. We might want to strengthen the checks for this if possible rather than just putting a bandaid on it for the test itself.

ahg-g commented on pkg/scheduler/apis/config/testing/defaults/defaults.go:
```diff
@@ -151,8 +151,41 @@ var PluginConfigsV1beta2 = []config.PluginConfig{
 }
 
-// PluginsV1beta3 default set of v1beta3 plugins.
+// PluginsV1beta3 is the set of default v1beta3 plugins (before MultiPoint expansion)
```
this will always be updated though, so I would replace "before" with "unrolled"

ahg-g replied:
this one not addressed yet

damemi replied:
@ahg-g could you clarify the wording you'd like here? Not sure I understand... `// PluginsV1beta3 is the set of default v1beta3 plugins (unrolled MultiPoint expansion)`?

ahg-g replied:
ok, perhaps I misunderstood this one, I guess the point is that those are after the multipoint is unrolled, if yes, then I don't think the original comment is correct, right?

damemi replied:
So `PluginsV1beta3` is the actual default, *before* the MultiPoint is unrolled (in other words, when all plugins are still just enabled in MultiPoint)

`ExpandedPluginsV1beta3` (below) is the default set of plugins *after* being unrolled.

So I think the original comment here is correct, if that makes sense?

ahg-g replied:
ok, sg

ahg-g commented on pkg/scheduler/framework/runtime/framework_test.go:
```diff
@@ -490,6 +494,204 @@ func TestNewFrameworkErrors(t *testing.T) {
+						{Name: scorePlugin1},
+						{Name: testPlugin},
+					},
+					Disabled: []config.Plugin{
```
this is needed to re-order, but if removed we will just override the weights, correct?

damemi replied:
It should, will add a test to confirm

damemi replied:
Added a test case. This needed an update because MultiPoint weights were overwriting individual Score weights (made Score to take precedence now)

While doing this I realized that we don't actually need to Disable any plugins to reorder them in individual points. Updated the test case to show that too

ahg-g commented on pkg/scheduler/framework/runtime/framework_test.go:
```diff
@@ -490,6 +494,204 @@ func TestNewFrameworkErrors(t *testing.T) {
+			},
+		},
+		{
+			name: "Multiple MultiPoint plugins",
```
typo?

ahg-g replied:
ah, multiple, I thought multipoint multipoint.

ahg-g commented on pkg/scheduler/framework/runtime/framework_test.go:
```diff
@@ -490,6 +494,204 @@ func TestNewFrameworkErrors(t *testing.T) {
+				MultiPoint: config.PluginSet{
+					Enabled: []config.Plugin{
+						{Name: testPlugin},
+						{Name: testPlugin},
```
this practically means we don't allow to override multiple point plugins (assuming the first one is from the default set), correct? I think we probably should allow it.

damemi replied:
I think that default plugins in `MultiPoint` can still be overridden via the same logic in the defaulting (`mergePlugins` ) that exists for all the other extension points. I'll add a test there to confirm though

damemi replied:
Added a test case (in `default_plugins_test.go`) to show this

alculquicondor commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -346,12 +330,85 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
 		}
 	}
 
+	// initialize MultiPoint plugins
```
just to be extra safe and ensure we are not changing the behavior for v1beta2:

Put this logic in a separate function and within that do at the beginning of the function:

```golang

if len(Plugins.Multipoint.Enabled) == 0 {
  return
}
```

damemi replied:
Made the change, but instead of putting the check in the function I wrapped the function call in the check to fail faster. Is there a preference to returning within the function call instead?

alculquicondor commented on pkg/scheduler/framework/runtime/framework_test.go:
```diff
@@ -490,6 +494,237 @@ func TestNewFrameworkErrors(t *testing.T) {
+			wantErr: "already registered",
+		},
+		{
+			name: "Allow overriding(reordering) MultiPoint/default without 'already registered' error",
```
isn't this already tested in `Override MultiPoint plugins weights`?

damemi replied:
True, this case is redundant. Removed

alculquicondor commented on pkg/scheduler/profile/profile.go:
```diff
@@ -86,20 +85,21 @@ type cfgValidator struct {
 	}
-	if len(cfg.Plugins.QueueSort.Enabled) != 1 {
-		return fmt.Errorf("one queue sort plugin required for profile with scheduler name %q", cfg.SchedulerName)
+	if len(f.ListPlugins().QueueSort.Enabled) != 1 {
```
Better validate this in `NewFramework`

damemi replied:
Do you mean to add an additional check in `NewFramework` for the same thing?

alculquicondor replied:
No, just remove this check here and do it in NewFramework.

damemi replied:
updated

alculquicondor commented on staging/src/k8s.io/kube-scheduler/config/v1beta3/types.go:
```diff
@@ -184,6 +184,16 @@ type Plugins struct {
+	// point the plugin has implemented. Disabling a plugin through MultiPoint disables that behavior.
+	// The same is true for disabling "*" through MultiPoint.
+	// Plugins can still be disabled through their individual extension points.
+	// Enabling a plugin through MultiPoint and a specific extension point produces an error
```
this is no longer true.

damemi replied:
Good catch, removed

ahg-g commented:
@damemi can you please not amend the changes, hard to review.

alculquicondor commented on pkg/scheduler/profile/profile.go:
```diff
@@ -95,9 +95,6 @@ func (v *cfgValidator) validate(cfg config.KubeSchedulerProfile, f framework.Fra
 	if v.m[f.ProfileName()] != nil {
 		return fmt.Errorf("duplicate profile with scheduler name %q", f.ProfileName())
 	}
-	if len(f.ListPlugins().QueueSort.Enabled) != 1 {
```
Now you can get rid of `f` as parameter :)

damemi replied:
The `validate` call has to come after framework instantiation so that we can check for expanded QueueSort plugins across all profiles.

Also for that reason, we still need the framework to be passed to this validate function, because the expansion doesn't actually modify the Profile that is passed to `NewFramework`. MultiPoint plugins are only expanded into the framework's internal representation of each extension.

alculquicondor replied:
Oh, we should have moved that check to `NewFramework` as well.

damemi replied:
`NewFramework` only handles one profile at a time, this check is making sure that there are no conflicting QueueSort plugins across all instantiated profiles.

alculquicondor replied:
Ahhh, I forgot about that.

LGTM then. @ahg-g anything else before squashing?

ahg-g commented on pkg/scheduler/framework/runtime/framework_test.go:
```diff
@@ -490,6 +494,200 @@ func TestNewFrameworkErrors(t *testing.T) {
+				},
+				PreScore: config.PluginSet{
+					Disabled: []config.Plugin{
+						{Name: "*"},
```
this is not quite relevant to the test case, right?

damemi replied:
true, I just added that so the output was a bit easier to parse. Removed the `"*"` cases to their own disabled extension point test case

ahg-g commented on pkg/scheduler/framework/runtime/framework_test.go:
```diff
@@ -490,6 +494,200 @@ func TestNewFrameworkErrors(t *testing.T) {
+				},
+				PreScore: config.PluginSet{
+					Disabled: []config.Plugin{
+						{Name: "*"},
```
is this relevant to the test case? I am guessing you are testing two things: reorder and disable an extension point.

damemi replied:
moved disable to its own case

ahg-g commented on pkg/scheduler/framework/runtime/framework_test.go:
```diff
@@ -490,6 +494,200 @@ func TestNewFrameworkErrors(t *testing.T) {
+				},
+				Score: config.PluginSet{
+					Enabled: []config.Plugin{
+						{Name: scorePlugin1},
```
what happens if you have a third multi point plugin that is not included here?

damemi replied:
The specific plugins take precedence over multipoint (the hierarchy is `Specific > MultiPoint > Default`). Added a 3rd plugin to show this in this example, but this is also kind of covered by the various `append` tests for `mergePlugins` (https://github.com/kubernetes/kubernetes/blob/bf2e6670f6c9a447b6c10170bf9bd5c47e4c0bf4/pkg/scheduler/apis/config/v1beta3/default_plugins_test.go#L373 for MultiPoint, but there are other cases for <v1beta3 defaults I believe)

ahg-g commented:
looks good to me other than the nit related to the comment.

ahg-g commented:
@liggitt this is ready for API review I think.

liggitt commented on staging/src/k8s.io/kube-scheduler/config/v1beta3/types.go:
```diff
@@ -184,6 +184,14 @@ type Plugins struct {
+
+	// MultiPoint is a simplified config section to enable plugins for all valid extension points.
+	// Plugins enabled through MultiPoint will automatically register for every individual extension
+	// point the plugin has implemented. Disabling a plugin through MultiPoint disables that behavior.
```
> Disabling a plugin through MultiPoint disables that behavior.

What does "that" refer to?

If you set `multiPoint.disabled = [{name:a}]`, does that disable `a` in all individual points, or just cancel any opinion `multiPoint` has about plugin `a`?

If you set `multiPoint.disabled = [{name:"*"}]`, does that disable all plugins in all individual points, or just cancel any opinion `multiPoint` has about all plugins?

damemi replied:
It depends on the full config. If you set `multiPoint.disabled = [{name:a}]` this cancels any opinion `multiPoint` has about plugin `a`. So, if `multiPoint` was the only place that plugin was configured, then it disables the plugin entirely. But it can still be enabled through individual extension points (because of the hierarchy, `Specific > MultiPoint > Default`.. if there is no `Specific` then `MultiPoint` is the highest precedence)

For `multiPoint.disabled = [{name:"*"}]`, it's the same behavior just applied to all plugins. This is essentially explicitly saying the config will only rely on Specific extension point configs

Huang-Wei replied:
> For multiPoint.disabled = [{name:"*"}], it's the same behavior just applied to all plugins. This is essentially explicitly saying the config will only rely on Specific extension point configs

I also a concern here. What if `multiPoint.disabled = [{name: "*"}]` is specified along with `multiPoint.enabled = [{name: "A"}]`? Does that mean: (1) plugin A should be enabled, or (2) all plugins disabled. BTW: suppose the regular extension point config is nil.

damemi replied:
Answered Wei's question above and added cases to cover it https://github.com/kubernetes/kubernetes/pull/105611#discussion_r749410464

alculquicondor commented on pkg/scheduler/apis/config/testing/defaults/defaults.go:
```diff
@@ -151,8 +151,41 @@ var PluginConfigsV1beta2 = []config.PluginConfig{
+		},
+	},
+	Score: config.PluginSet{
+		Disabled: []config.Plugin{
+			{Name: names.VolumeBinding},
+		},
```
I can see how fiddling with the configuration is confusing. Instead, what we can do is make the VolumeBinding Score implementation be no-op (0 score for every node).

damemi replied:
This is a translation of the existing implementation to the new defaults as MultiPoint plugins. Previously, the VolumeBinding Score plugin was implicitly disabled by simply not being included. Now that we register default plugins through multipoint, that enables VolumeBinding/Score by default.

The old code would read the featuregate flag and update the default config as needed to enable the Score plugin. This does the inverse, and once the featuregate is enabled by default this code will be removed.

damemi commented on staging/src/k8s.io/kube-scheduler/config/v1beta3/types.go:
```diff
@@ -184,6 +184,14 @@ type Plugins struct {
+	// point the plugin has implemented. Disabling a plugin through MultiPoint disables that behavior.
+	// The same is true for disabling "*" through MultiPoint.
+	// Plugins can still be disabled through their individual extension points.
+	// Plugins configured through specific extension points take precedence and run first over MultiPoint by default.
```
All of these configs are valid, and we have test coverage for them. The basic hierarchy is `Specific Point > MultiPoint > Default`, which preserves the current behavior for existing (v1beta2) configs. So a v1beta2 config, which will be using Specific points only, has the same output as a v1beta3 config. The docs for MultiPoint is started here, which covers some of these cases: https://github.com/kubernetes/website/pull/30442

alculquicondor replied:
I think we still need more clarity on this godoc comment.
I don't think there is a difference between disabling by name or by `*`, but we still need to clarify all the multipoint:(enabled/disabled), specific(enabled/disabled) pairs.

damemi replied:
I tried to expand this godoc a bit more, I think that explaining the hierarchy of extension points and disabled/enabled logically covers the different combination patterns without needing to individually call out each one (but I also included a couple examples too).

I could add as much as we want for this comment. We are going to have a much more detailed docs section with the PR I linked which is a better format for covering all specific edge cases imo, so keeping that in mind I don't know if we want this to get too long

alculquicondor commented on pkg/scheduler/apis/config/v1beta3/default_plugins.go:
```diff
@@ -29,91 +29,36 @@ import (
-		},
-		Filter: v1beta3.PluginSet{
-			Enabled: []v1beta3.Plugin{
+				{Name: names.NodeAffinity, Weight: pointer.Int32(2)},
```
Only Filter has a meaningful order, as the scheduler breaks after the first Filter failure. Other extension points have order (like PostFilter or Bind), but in kube-scheduler we only have one plugin for each.

Please make sure we are preserving the same order of the Filter plugins.

damemi replied:
We could reorder these if that's the case @ahg-g wdyt?

ahg-g replied:
yeah, the order here should match the one we used to have in Filter

damemi replied:
Order has been reestablished

damemi commented on pkg/scheduler/apis/config/v1beta3/default_plugins.go:
```diff
@@ -122,17 +67,17 @@ func getDefaultPlugins() *v1beta3.Plugins {
-		config.Score.Enabled = append(config.Score.Enabled, v1beta3.Plugin{Name: names.VolumeBinding, Weight: pointer.Int32Ptr(1)})
+		// Remove VolumeBinding from Score.Disabled if the feature gate is enabled
+		// Since VolumeBinding is the only disabled plugin right now in the default registry, we can just set Disabled to nil
+		config.Score.Disabled = nil
```
Answered above, but this is just inverting the current logic since MultiPoint enables every extension by default.

liggitt replied:
it looks like the VolumeBinding plugin already has a no-op scorer if the feature is off, which makes its Score function inert... not sure why we need to mess with the config here based on feature gates

damemi replied:
@liggitt that's a good point. In that case, I don't know why we were originally doing the logic to modify the default config here. Looks like @ahg-g added that with `applyFeatureGates` in https://github.com/kubernetes/kubernetes/commit/265ef1741f951290f2b302d81c694186b79decfa?

I do think it is beneficial to clearly show that the plugin is disabled, specifically when logging the internal config. Otherwise, it will show as enabled but it's really doing nothing. That sounds weird to me

ahg-g replied:
yeah, we can just depend on the logic in the plugin itself.

ahg-g replied:
@damemi we want admins to think about features, less about plugins. I am ok with removing the logic here and depend on the one in the plugin itself.

damemi replied:
Should we do the same for `DefaultPodTopologySpread` and just remove `applyFeatureGates()` entirely?

alculquicondor replied:
I think we can leave that for now. I'm planing to GA next release. At least there are no partial disables in that case.

damemi replied:
I don't think a partial or full disable is relevant, we're going to be disabling some plugins silently based on a featuregate and others explicitly through the config. That's a confusing inconsistency

alculquicondor replied:
I'm less worried about enabling a plugin based on a feature gate, when the feature gate refers to the entire plugin.

In the next release, when the feature graduates to GA, the plugin would disappear from defaults. But the plugin should still be available for use. Can we do that without bumping up the version? I think removing a plugin from the defaults is backwards compatible.

ahg-g replied:
I don't think consistency is important in this context.

VolumeBinding scoring is an extension of the plugin, we don't want to encourage thinking about disabling/enabling extensions, we want to shift towards whole feature/plugin model. To make that point, I don't think we should be disabling the extension here based on the feature gate.

Enabling SelectorSpread is different in my opinion because it is about a whole plugin; so module the comment I left there, I think we can keep it as is (enabled/disabled based on the feature gate).

damemi replied:
Thought about this some more, and I think the distinction @ahg-g and @alculquicondor are making is reasonable. The goal of MultiPoint is to abstract away the specific extension points as much as possible from administrators.

While I do think it is confusing to see a plugin "enabled" for one extension point where it's really a no-op, I am also biased toward thinking in terms of the specific extension points because that's all we have had until this point.

So it appears inconsistent to treat these different plugins differently, but like Abdullah mentioned the difference is in an entire feature, not an extension point.

Going in the other direction for that, there could be features that are subsets of specific extension points. (For example, if Scoring was enabled but there was a gate to tune the scoring method even further). That type of setting doesn't show up in the Component Config either, but it is still valid to say the plugin is enabled. We want to apply the same idea to MultiPoint, where entire plugins are the features.

So, I'll cede this and remove the logic that I had in this function. @Huang-Wei what do you think?

Huang-Wei replied:
My major concern was the ambiguousness it (the inconsistency) may deliver to the developers - they will have to understand the underlying nuances to better contribute to the codebase.

While for administrators, this approach is fine with me. For non-GA features, they use the feature gate to control  its on/off without worrying about whether this feature is part of a multipoint or not; for GAed features, they leverage the `enabled/disabled` directives or tune plugin arguments of multipoint or regular point to turn on/off the feature.

alculquicondor commented on staging/src/k8s.io/kube-scheduler/config/v1beta3/types.go:
```diff
@@ -187,6 +187,9 @@ type Plugins struct {
 	// MultiPoint is a simplified config section to enable plugins for all valid extension points
+	// Plugins enabled through MultiPoint will automatically register for every individual extension point the plugin has implemented.
+	// Disabling a plugin through MultiPoint disables that behavior. The same is true for disabling "*" through MultiPoint.
+	// Plugins can still be disabled through their individual extension points.
```
btw, this is not longer true.

damemi commented on staging/src/k8s.io/kube-scheduler/config/v1beta2/types.go:
```diff
@@ -191,6 +191,9 @@ type Plugins struct {
 	// PostBind is a list of plugins that should be invoked after a pod is successfully bound.
 	PostBind PluginSet `json:"postBind,omitempty"`
+
+	// MultiPoint is a simplified config section to enable plugins for all valid extension points.
```
Docs PR for multipoint is here: https://github.com/kubernetes/website/pull/30442

Huang-Wei reviewed (commented):
Haven't looked into tests yet. I'd like to see tests to clarify the interaction with enable/disable out-of-tree plugins via MultiPoint.

ahg-g commented on pkg/scheduler/apis/config/v1beta3/default_plugins.go:
```diff
@@ -122,17 +67,23 @@ func getDefaultPlugins() *v1beta3.Plugins {
+		for index, plugin := range config.MultiPoint.Enabled {
+			if plugin.Name == names.PodTopologySpread {
+				config.MultiPoint.Enabled[index] = s
+				config.MultiPoint.Disabled = append(config.MultiPoint.Disabled, v1beta3.Plugin{Name: names.PodTopologySpread})
```
I don't understand this, why are we disabling the whole PodTopologySpread plugin, that doesn't sound right. We only need to enable `SelectorSpread`

damemi replied:
This was something I noticed while reordering the Filter plugins above, and maybe I misunderstood it. But if the `DefaultPodTopologySpread` feature gate is disabled, shouldn't that disable PodTopologySpread as a default plugin as well? Looking at it again, we weren't doing that before so maybe that's not the intent. I can revert this part if so

ahg-g replied:
We need to revert it, when this feature is enabled, we should simply add the plugin. We don't want to completely disable `PodTopologySpread`, we only want to skip default spreading via PodTopologySpread which is done here https://github.com/kubernetes/kubernetes/blob/9b45983d3c8c246948682e678caa721b6d0bed0c/pkg/scheduler/apis/config/v1beta3/defaults.go#L230

damemi replied:
Okay, will do then

damemi commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -384,6 +396,86 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
+			disabledPlugins.Insert(disabledPlugin.Name)
+		}
+
+		for _, ep := range profile.Plugins.MultiPoint.Enabled {
```
I added some cases to cover this, but it is essentially the same logic as this would have applied with `"*"` for default plugins in the old config. That is, specifically enabling a plugin for any extension point will enable it (including MultiPoint). 

We use the same function `mergePlugins()` which handles `"*"` for each extension point to disable all default plugins for that point, but after that it merges in any explicitly enabled plugins. This is the existing behavior

damemi commented:
Pretty sure I've gotten to all the outstanding comments at this point, if there's anything I missed please let me know

ahg-g commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -384,6 +423,24 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
+		// a weight of zero is not permitted, plugins can be disabled explicitly
+		// when configured.
+		f.scorePluginWeight[e.Name] = int(e.Weight)
+		if f.scorePluginWeight[e.Name] == 0 {
```
can you pls open an issue to track this, it is dangerous to leave it like that.

damemi replied:
opened in https://github.com/kubernetes/kubernetes/issues/106426

ahg-g commented:
/milestone v1.23

code freeze is tomorrow, @liggitt I think the comments has been addressed, please take a look.

liggitt commented:
ack, will look this afternoon

damemi commented on cmd/kube-scheduler/app/server_test.go:
```diff
@@ -379,7 +432,7 @@ leaderElection:
 				}
 
-				if diff := cmp.Diff(tc.wantPlugins, gotPlugins); diff != "" {
+				if diff := cmp.Diff(tc.wantPlugins, gotPlugins, cmpopts.SortSlices(func(x, y config.Plugin) bool { return strings.Compare(x.Name, y.Name) > 0 })); diff != "" {
```
Nope, just had to update the test structs to reflect the ordering as passed by the reordered filter plugins

damemi commented on cmd/kube-scheduler/app/server_test.go:
```diff
@@ -188,15 +196,60 @@ leaderElection:
-						PreBind:    defaults.PluginsV1beta3.PreBind,
-						Reserve:    defaults.PluginsV1beta3.Reserve,
+						QueueSort: defaults.ExpandedPluginsV1beta3.QueueSort,
+						PreFilter: config.PluginSet{
```
removed

ahg-g commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -346,12 +330,39 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
+
+	var totalPriority int64
+	if len(profile.Plugins.MultiPoint.Enabled) > 0 {
+		err, multiPointWeights := getScoreWeights(f, totalPriority, profile.Plugins.MultiPoint.Enabled)
```
I don't understand why we need to do this here, can't we just apply the default weights once after calculating the final set of score plugins?

damemi replied:
Ah that's a good point, yeah we can just work with the expanded Score weights. Ignore my comment above then :)

damemi replied:
Actually, we can't use the expanded score weights, because we don't actually expand the weights. This function does not modify the KubeSchedulerConfig. We only build `f.scorePlugins` from the config. Then, we get the weights from the config to translate it to `f.scorePluginWeight` later on. So we still need to look at both the MultiPoint weights and the individual Score weights as I mentioned in https://github.com/kubernetes/kubernetes/pull/105611/files#r750365051

Huang-Wei commented on pkg/scheduler/apis/config/v1beta3/default_plugins.go:
```diff
@@ -166,9 +100,15 @@ func mergePluginSet(defaultPluginSet, customPluginSet v1beta3.PluginSet) v1beta3
 	enabledCustomPlugins := make(map[string]pluginIndex)
 	// replacedPluginIndex is a set of index of plugins, which have replaced the default plugins.
 	replacedPluginIndex := sets.NewInt()
+	var disabled []v1beta3.Plugin
```
BTW: how about renaming it to `disabledPlugins`, and rename current `disabledPlugins` to `disabledPluginNames`?

ahg-g replied:
Good point, in the past we didn't care about disabled plugins in the default set because ot held no meaning really since one can just remove it from the enabled set. With multipoint we could give it a meaning.

damemi replied:
There are no disabled plugins in the default plugin set, the only one we had was VolumeBinding. So, I removed the logic to handle that in https://github.com/kubernetes/kubernetes/pull/105611/commits/8a81833da8c9a48b65ced41a00ad8289cccf7f63#diff-f1c8947cb8d7c33e8b99363e4002b12fbdf0bed7fb11eb857529204469b9e769L111-L112

I can re-add the logic in case we do add disabled plugins in the future, but for now it would do nothing.

There is a unit test for disabling a default plugin manually: https://github.com/kubernetes/kubernetes/blob/a11794e34823bfd6fd1159261cda18a1335cbd9f/pkg/scheduler/apis/config/v1beta3/default_plugins_test.go#L390

ahg-g replied:
ok, since this logic is versioned, we can keep it as is because there are no disabled extensions in the default set of plugins. Changing the default set requires upping the version, and only then we need to handle this if we ever wanted to. 

I am ok with both adding it back or keeping it as is, we are safe in both cases.

damemi replied:
I re-added the logic to handle this with a test case just in case we do add default disabled plugins in the future.

Huang-Wei replied:
> I re-added the logic to handle this with a test case just in case we do add default disabled plugins in the future.

Thanks. The latest code looks good.

Huang-Wei commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -346,12 +330,39 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
 	}
 
+	// initialize multiPoint plugins to their expanded extension points
+	if len(profile.Plugins.MultiPoint.Enabled) > 0 {
```
Why add this condition? What if the multiPoint only carries `disabled`? Given the following pluginSet:

```go
			customPlugins: &v1beta3.Plugins{
				MultiPoint: v1beta3.PluginSet{
					Disabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin1"},
						{Name: "DefaultPlugin2"},
					},
				},
			},
			defaultPlugins: &v1beta3.Plugins{
				Filter: v1beta3.PluginSet{
					Enabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin1"},
					},
				},
				Score: v1beta3.PluginSet{
					Enabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin2"},
					},
				},
			},
```

The merged pluginSet would be:

```go
				Filter: v1beta3.PluginSet{
					Enabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin1"},
					},
				},
				Score: v1beta3.PluginSet{
					Enabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin2"},
					},
				},
				MultiPoint: v1beta3.PluginSet{
					Disabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin1"},
						{Name: "DefaultPlugin2"},
					},
				},
```

In this case, we should still handle MultiPoint even if it only has `disabled`, right? Because without it, the result would carry regular extension point **only** - which is processed earlier in updatePluginList().

damemi replied:
MultiPoint only takes effect if it has Enabled plugins. Otherwise, there will be no MultiPoint expansion. Putting plugins into `multiPoint.Disabled` without `multiPoint.Enabled` doesn't do anything. As you showed in your example, those plugins are specifically enabled through individual extension points.

ahg-g replied:
can we resolve this one?

Huang-Wei replied:
Bump. (`mergePluginSet()` in default_plugins.go looks good now)

And make sure to have UTs covering this.

damemi replied:
This was covered by other tests, the `disabled` field is carried in. I added a new test specifically showing this without any other settings.

ahg-g replied:
sg, @Huang-Wei  good to resolve this comment?

Huang-Wei replied:
A corner case here: what if pluginA is enabled AND also disabled in multipoint? In the case below, we should disable scorePlugin1, right? (maybe it's not a _corner_ case, as the plugins here is merged from default plus user specified ones - for example, default plugins enabled scorePlugin1, but user disables it via CC)

```go
				MultiPoint: config.PluginSet{
					Enabled: []config.Plugin{
						{Name: queueSortPlugin},
						{Name: bindPlugin},
						{Name: scorePlugin1},
					},
					Disabled: []config.Plugin{
						{Name: scorePlugin1},
					},
				},
```

damemi replied:
Isn't that this test? https://github.com/kubernetes/kubernetes/pull/105611/files#diff-392546fd01ec251b85f84db8e534ad5d22aa4e58de40b629417d0af0f90916b9R390-R432

damemi replied:
> for example, default plugins enabled scorePlugin1, but user disables it via CC)

that's the exact test case I shared. The user disabling it takes precedence as defined by the hierarchy of `Specific > MultiPoint > Default(MultiPoint)`

Huang-Wei replied:
My point is that scorePlugin1 should be disabled in this case.

damemi replied:
It is disabled, the user disabling it via CC takes precedence. See that in that test, the user disables `DefaultPlugin2` via CC, and the expected output shows `DefaultPlugin2` disabled. This is the same scenario.

Huang-Wei replied:
In other words, in terms of multiPoint, we can discard `disabled: *`, but not for `disabled: <specific plugin>`.

damemi replied:
No, both of those are valid in certain use cases and there are test cases for both.

Huang-Wei replied:
> No, both of those are valid in certain use cases and there are test cases for both.

Without logic of dealing with `multipoint.disabled`, how can you achieve that? Am I missing something?

Also, the `scorePlugin1` is still there in the wantPlugins, in subtest "disable and enable MultiPoint plugin by name".

damemi replied:
> Without logic of dealing with multipoint.disabled, how can you achieve that? Am I missing something?

`mergePlugins()` handles this when the component config is processed. If the plugin is disabled, it won't be added to the merged config that gets processed here. There is no need for us to handle `multiPoint.disabled` at this point because it has already been merged.

> Also, the scorePlugin1 is still there in the wantPlugins, in subtest "disable and enable MultiPoint plugin by name".

The difference in that test is that the user *enabled* that plugin as well. Enabling always takes precedence over disabling. This allows for overriding and is consistent with the current behavior.

damemi replied:
See comments in the godoc: https://github.com/kubernetes/kubernetes/pull/105611/files#diff-5aed18dafd438aa779df20e2474aeeaa6a75ef70493c282b4c428acf90553382R200-R202

damemi replied:
The reason we only need to act in the framework runtime if `len(multiPoint.enabled) > 0` is because there will be nothing to do if there are no plugins enabled through multiPoint. If there were (either custom or default), they will have been merged into the single config here. We do check some of the disabled plugins in this section to make sure the plugin isn't specifically disabled by name. But if there are no `multiPoint` plugins enabled, then we don't care about `multiPoint.disabled`.

damemi replied:
`multiPoint.disabled` isn't just shorthand to disable every plugin in every extension point. It is only for disabling the multiPoint behavior, ie the expanding into those extension points

Huang-Wei replied:
> `mergePlugins()` handles this when the component config is processed. If the plugin is disabled, it won't be added to the merged config that gets processed here.

> If there were disabled plugins in multiPoint (either custom or default), they will have been merged into the single config here.

This makes sense to me. Thanks for the explanation.

Huang-Wei commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -384,6 +395,91 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
+		}
+
+		disabledSet := sets.NewString()
+		for _, disabledPlugin := range e.plugins.Disabled {
+			disabledSet.Insert(disabledPlugin.Name)
+		}
```
I suppose we're not covering the logic properly. See my another comment, the disabledSet seems to have dropped default pluginSet's `disabled`, and thus the data here is incorrect.

damemi replied:
I commented on that above, but we don't have any default plugins that are disabled. I can re-add [the logic](https://github.com/kubernetes/kubernetes/commit/8a81833da8c9a48b65ced41a00ad8289cccf7f63#) to handle them just in case we add any in the future.

If we do that, this will then handle those cases as well. That's because the default plugins and custom config are merged at this point into a single config.

ahg-g replied:
@damemi  we shouldn't make assumptions about the versioned logic. The implementation here should handle potential changes in future versions.

damemi replied:
Re-added the logic to handle default disabled plugins

Huang-Wei replied:
Yeap, we shouldn't make the assumptions as other consumers like ClusterAutoscaler may instantiate the plugins any way they like.

ahg-g commented on pkg/scheduler/apis/config/testing/defaults/defaults.go:
```diff
@@ -151,22 +151,50 @@ var PluginConfigsV1beta2 = []config.PluginConfig{
 	PreFilter: config.PluginSet{
 		Enabled: []config.Plugin{
-			{Name: names.NodeResourcesFit},
+			{Name: names.NodeAffinity},
 			{Name: names.NodePorts},
+			{Name: names.NodeResourcesFit},
 			{Name: names.VolumeRestrictions},
+			{Name: names.VolumeBinding},
```
it is a byproduct of the multipoint order which is now based on filter.

ahg-g replied:
Yes.

ahg-g commented on pkg/scheduler/apis/config/testing/defaults/defaults.go:
```diff
@@ -195,32 +223,33 @@ var PluginsV1beta3 = &config.Plugins{
-			// Weight is doubled because:
-			// - This is a score coming from user preference.
-			{Name: names.InterPodAffinity, Weight: 2},
+			{Name: names.TaintToleration, Weight: 3},
```
it is a byproduct of the multipoint order which is now based on filter.

liggitt commented on pkg/scheduler/profile/profile.go:
```diff
@@ -55,13 +54,13 @@ func NewMap(cfgs []config.KubeSchedulerProfile, r frameworkruntime.Registry, rec
 	v := cfgValidator{m: m}
 
 	for _, cfg := range cfgs {
-		if err := v.validate(cfg); err != nil {
```
moving validate after newProfile looks weird… is newProfile mutating cfg now?

damemi replied:
> is newProfile mutating cfg now?

Yes, the multiPoint expansion logic occurs at the instantiation of the framework. This happens here because we need to try casting each MultiPoint plugin to every other plugin type per the design we settled on. This is done on individual Profiles.

After that expansion, the `validate()` call does various cross-Profile checks to ensure the entire config is valid (for example, no conflicting QueueSort plugins are enabled across Profiles). 

As part of this, some checks from `validate()` actually moved in to `newProfile` to still fail at the expected time. There are tests for these errors that we made sure to keep in tact

damemi replied:
> > is newProfile mutating cfg now?

> Yes,

Sorry, to be more clear this isn't *modifying* the cfg. It's expanding cfg into a framework. The rest of my comment still holds, we are just doing some of these checks on the framework to avoid modifying the cfg

damemi commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -346,12 +330,39 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
+		if err != nil {
+			return nil, err
+		}
+		totalPriority += multiPointWeights
```
The reasoning here is that we're still doing checking in this function to make sure the total score doesn't exceed the max value. Since we could have different score plugins configured in both MultiPoint and Score, we need to consider both

ahg-g replied:
wouldn't that consider the weight for the same plugin twice if it was overridden?

damemi replied:
True, so we could handle that by passing another set between the two calls. I'll add that.

Unfortunately we can't just use the final expanded weights without refactoring how the framework stores Score plugins. That's not a huge refactor, but maybe a better todo

damemi replied:
@ahg-g okay, solved this with the changes in https://github.com/kubernetes/kubernetes/pull/105611/commits/a517726dae02b9a82073391489f0e7d7b1eaa8a9

The pluginWeights are a map, so we don't need to worry about double-calculation. Combine it into one call by appending MultiPoint plugins to Score plugins. If the plugin already exists, assume that's because it was present earlier in the list as a Score plugin (don't know how we want to handle specifying the same plugin multiple times with different weights... I think that's user error). Check that each plugin actually implements `Score` before calculating the weight.

ahg-g replied:
thanks, you can resolve this comment.

liggitt commented:
It looks like there’s still gaps (or at least concerns, based on @Huang-Wei’s comments) about some of the config merging bits… those need to be agreed on and demonstrated in tests to make sure the combinations are understood and correct

ahg-g commented:
@damemi can you please close the comments that have already been addressed so we can focus on the ones that are still actually open?

damemi commented on cmd/kube-scheduler/app/server_test.go:
```diff
@@ -84,12 +84,18 @@ users:
 	pluginConfigFile := filepath.Join(tmpDir, "plugin.yaml")
 	if err := ioutil.WriteFile(pluginConfigFile, []byte(fmt.Sprintf(`
-apiVersion: kubescheduler.config.k8s.io/v1beta2
+apiVersion: kubescheduler.config.k8s.io/v1beta3
```
v1beta2 is also getting a multiPoint field to handle round-tripping through the internal version, which is what all the logic operates on. So, this shouldn't make a difference right?

liggitt replied:
It matters if it changes behavior of existing v1beta2 config files. Changing this here makes it hard to tell if that happened

damemi replied:
It shouldn't, but I split this out into a v1beta3 test and kept the old v1beta2 test.

The v1beta3 test looks different now because I specifically added MultiPoint plugins to achieve the same output as it would have been without multipoint (ie, v1beta2' s output). But the semantics for `Disabled`/`Enabled` in individual extension points is the same, and individual points take precedence, so even though default plugins are now "MultiPoint" the v1beta2 test doesn't break

damemi commented:
Pushed the small changes as well as re-adding the logic to handle disabled plugins in the default registry. We don't have any of these right now, but the logic is there with a test case in case we add any in the future.

Also went through and resolved as many conversations as I could as @liggitt requested.

ahg-g commented on pkg/scheduler/apis/config/v1beta3/default_plugins_test.go:
```diff
@@ -454,6 +454,37 @@ func TestMergePlugins(t *testing.T) {
+					Enabled: []v1beta3.Plugin{
+						{Name: "DefaultPlugin"},
+					},
+					Disabled: []v1beta3.Plugin{
```
the more important case is where you disable an extension, not the whole plugin.

damemi replied:
That already has its own test case above https://github.com/kubernetes/kubernetes/pull/105611/files#diff-392546fd01ec251b85f84db8e534ad5d22aa4e58de40b629417d0af0f90916b9R390-R432

ahg-g replied:
that test doesn't cover the case where an extension point is disabled in the default set.

damemi replied:
Ah, I see, misread that. Added

ahg-g commented:
I have one final comment: https://github.com/kubernetes/kubernetes/pull/105611#discussion_r750513677

otherwise this lgtm

Huang-Wei commented on pkg/scheduler/apis/config/v1beta3/default_plugins_test.go:
```diff
@@ -451,6 +359,132 @@ func TestMergePlugins(t *testing.T) {
+		},
 	}
 
 	for _, test := range tests {
```
(this is pre-existing)

Could you help update L492~L493 to:

```go
			gotPlugins := mergePlugins(test.defaultPlugins, test.customPlugins)
			if d := cmp.Diff(test.expectedPlugins, gotPlugins); d != "" {
				t.Fatalf("plugins mismatch (-want +got):\n%s", d)
			}
```

(setting the plugins back to subtest looks weird)

damemi replied:
Done

damemi commented:
Got to the last comments. If we want to do a final round, should I squash this now? Don't want to preemptively squash to mess anyone up

ahg-g commented:
> Got to the last comments. If we want to do a final round, should I squash this now? Don't want to preemptively squash to mess anyone up

there is still this one open/not addressed: https://github.com/kubernetes/kubernetes/pull/105611#discussion_r749762361

damemi commented:
> > Got to the last comments. If we want to do a final round, should I squash this now? Don't want to preemptively squash to mess anyone up
> 
> there is still this one open/not addressed: [#105611 (comment)](https://github.com/kubernetes/kubernetes/pull/105611#discussion_r749762361)

That is covered by the test in https://github.com/kubernetes/kubernetes/pull/105611/files#diff-392546fd01ec251b85f84db8e534ad5d22aa4e58de40b629417d0af0f90916b9R390-R432. This test sets `multiPoint.disabled` and shows that it is carried through to the merged config. Whether or not the config only has `multiPoint.disabled` is irrelevant, because without any plugins specifically enabled in multipoint the entire field is a no-op.

Still, I added another test specifically for the custom config to only have `multiPoint.disabled` set and show that it's merged just as every other point. I also squashed at this point

ahg-g commented:
/lgtm

Huang-Wei commented:
https://github.com/kubernetes/kubernetes/pull/105611#discussion_r750643122 was folded. So re-comment here, I'd like to see subtest "disable and enable MultiPoint plugin by name" in pkg/scheduler/framework/runtime/framework_test.go to show the wantPlugins as:

```go
			wantPlugins: &config.Plugins{
				QueueSort: config.PluginSet{Enabled: []config.Plugin{{Name: queueSortPlugin}}},
				PreScore: config.PluginSet{Enabled: []config.Plugin{
					{Name: scorePlugin1},
				}},
				Bind: config.PluginSet{Enabled: []config.Plugin{{Name: bindPlugin}}},
			},
```

ahg-g commented:
> [#105611 (comment)](https://github.com/kubernetes/kubernetes/pull/105611#discussion_r750643122) was folded. So re-comment here, I'd like to see subtest "disable and enable MultiPoint plugin by name" in pkg/scheduler/framework/runtime/framework_test.go to show the wantPlugins as:
> 
> ```go
> 			wantPlugins: &config.Plugins{
> 				QueueSort: config.PluginSet{Enabled: []config.Plugin{{Name: queueSortPlugin}}},
> 				PreScore: config.PluginSet{Enabled: []config.Plugin{
> 					{Name: scorePlugin1},
> 				}},
> 				Bind: config.PluginSet{Enabled: []config.Plugin{{Name: bindPlugin}}},
> 			},
> ```

The current test is correct. The disabled set of multipoint plugins has no meaning at the point when trying to expand multipoint enabled plugins, this is what "disable and enable MultiPoint plugin by name" is showing.

Huang-Wei commented:
https://github.com/kubernetes/kubernetes/pull/105611#discussion_r750654651 resolves my concern.

/lgtm

ahg-g commented:
@liggitt we converged :)

liggitt commented:
/approve

for config API bits
